### PR TITLE
feat: codegen for hyper

### LINKED ISSUE #149: Add CodeGen for Rust (hyper)
**Tell us about the task you want to perform and are unable to do so because the feature is not available**

Add CodeGen for Rust
Library - https://hyper.rs/

Guidelines for adding a codegen - https://github.com/foss42/apidash/discussions/80

### PR DESCRIPTION
**Pull request by Clasherzz**
## PR Description
Through this PR I have added codegen file for rust hyper library and made necessary changes in other files to accomadate the change. The cargo.toml file is attached below:

```[package]
name = "reqwest"
version = "0.1.0"
edition = "2021"

[dependencies]
reqwest = { version = "0.12.7", features = ["multipart", "rustls-tls"] }

hyper = { version = "0.14", features = ["client", "http1"] }
hyper-tls = "0.5"
hyper-rustls = "0.22"

bytes = "1.2"

serde_urlencoded = "0.7" 

# Tokio for async runtime
tokio = { version = "1", features = ["full"] }

# Multipart for handling form data
multipart = "0.18"

# Serde for serializing/deserializing JSON
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
```

NOTE: Hyper doesnt support multipart and formdata. To accomplish this I have used reqwest library (I have referred reqwest codegen for this) but the my package version seems to conflict . But the implementation is exactly the same as that in codege of Reqwest and chances of error are less. I have contacted @Tanish2002 , the author of reqwest codegen file asking for his cargo.toml file but hasn't received a response . I will send the updated cargo.toml once I receive that.

## Related Issues

- Related Issue #149
- Closes #149

### Checklist
- [x] I have gone through the [contributing guide](https://github.com/foss42/apidash/blob/main/CONTRIBUTING.md)
- [x] I have updated my branch and synced it with project `main` branch before making this PR
- [x] I have run the tests (`flutter test`) and all tests are passing

## Added/updated tests?
- [x] Yes
- [x] Yes, i have updated the tests

##Preview
Screenshot 2024-09-20 203319
Screenshot 2024-09-20 203334
Screenshot 2024-09-20 203356
Screenshot 2024-09-20 203556

### CODE REVIEW

**Comment by Tanish2002** [comment]
@Clasherzz I think using 2 libraries (reqwest and hyper) that do the same thing in the codegen is counterproductive. 

Since hyper doesn't support multipart forms you can create the multipart data yourself. APIDash does expose the boundary data of attaced files.

https://github.com/foss42/apidash/pull/190 might help you a ton. ureq didn't support multipart forms as well so I did a custom implementation.

[python's http_client](
https://github.com/foss42/apidash/blob/main/lib/codegen/python/http_client.dart) has the same implementation as well

**Comment by Clasherzz** [comment]
Thanks for the quick response Tanish . I was informed to use another library by @ashitaprasad . Which is why opted for reqwest . Let me see if i can create a multipart .

**Comment by ashitaprasad** [comment]
@Clasherzz By using another library, I meant what we have done for Python requests where we use `requests_toolbelt` for Multipart encoding 
https://github.com/foss42/apidash/blob/main/lib/codegen/python/requests.dart
You have to use a helper library for multipart that is commonly used by the people using hyper. (Check with the project community or raise an issue asking if no community is available)

`reqwest` cannot be used as it is a competing HTTP client library and it is a high probability people will not use them together.

**Comment by Clasherzz** [comment]
A custom implementation just like in #190 would also be accepted right?

**Comment by ashitaprasad** [comment]
No @Clasherzz 
Custom implementation are no longer accepted and we have migrated from it for some earlier implementations.
You have to pick the library which is most used by the hyper community as that is what people who are generating the code will prefer.

There are a lot of crates available for this very purpose.

Some starting points:
- https://crates.io/keywords/multipart
- https://crates.io/search?q=form%20data

You can further investigate.

**Comment by Clasherzz** [comment]
I haven't worked much with boundaries and is still a little confused. Here i was trying to use hyper_multipart_rfc7578 and to  my knowledge the boundary should be generated by its own. Upon execution of :
```
extern crate hyper_multipart_rfc7578 as hyper_multipart;
extern crate hyper_tls;

use hyper::{Body, Client, Request};
use hyper_tls::HttpsConnector;
use hyper_multipart::client::multipart;
use hyper::body::to_bytes;
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // Create an HTTPS connector
    let https = HttpsConnector::new();
    let client = Client::builder().build::<_, Body>(https);

    // Create a multipart form and add text fields
    let mut form = multipart::Form::default();
    form.add_text("times", "3");
    form.add_text("text", "API");
    form.add_text("sep", "|");

    // Build the request
    let req_builder = Request::post("https://api.apidash.dev/io/form");
    
  //  let req_builder = Request::post("http://localhost:3000/api/hello");
    let req = form.set_body_convert::<Body, multipart::Body>(req_builder)
        .map_err(|e| format!("Failed to set body: {}", e))?;

    // Send the request and await the response
    match client.request(req).await {
        Ok(res) => {
            // Print the response status
            println!("Response Status: {}", res.status());

            // Convert the response body to bytes and print it as a string
            let body_bytes = to_bytes(res).await?;
            let body_string = String::from_utf8(body_bytes.to_vec())?;
            println!("Response Body: {}", body_string);
        }
        Err(e) => {
            eprintln!("Failed to send request: {}", e);
        }
    }

    Ok(())
}
```
An Error was  generated:
```
Failed to send request: error trying to connect: A required certificate is not within its validity period when verifying against the current system clock or the timestamp in the signed file. (os error -2146762495)
```
Does this mean i have to send a custom boundary through a header. The same worked fine when i send it a local endpoint .Could you point out what going wrong.

```

**Comment by ashitaprasad** [comment]
@Clasherzz go through some example for text and file multipart input.
The whole purpose of using an external library is make it easier as user does not need not deal with creation of boundaries and content.

**Comment by Clasherzz** [comment]
I am aware of this testcase and my previous code made a request  towards the same endpoint  (https://api.apidash.dev/io/form) . I thought this was the error of my code but now that you mentioned i ran the exact code in this test case :
```
import requests
from requests_toolbelt.multipart.encoder import MultipartEncoder

url = 'https://api.apidash.dev/io/form'

payload = MultipartEncoder({
  "text": "API",
  "sep": "|",
  "times": "3",
})

headers = {
  "content-type": payload.content_type
}

response = requests.post(url, data=payload, headers=headers)

print('Status Code:', response.status_code)
print('Response Body:', response.text)
```
But its showing an SSLCertification Error . I did this same for http.client package which resulted in similar error.
I am starting to think this is more of a server side problem .

**Comment by ashitaprasad** [comment]
@Clasherzz The server-side issue is now resolved. Please check again.

**Comment by Clasherzz** [comment]
The branch is now ready for review 
The new cargo.toml would be useful for testing :
```
[package]
name = "hypermulti"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
hyper = { version = "0.14", features = ["client", "http1"] }
hyper-tls = "0.5"
hyper-rustls = "0.22"


bytes = "1.2"

serde_urlencoded = "0.7" 
tokio = { version = "1", features = ["full"] }

hyper-multipart-rfc7578 = "0.7"


serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"

```

**Comment by ashitaprasad** [comment]
@Clasherzz You have to include the test file for this codegen. You can refer to the test files for other codegens here -
https://github.com/foss42/apidash/tree/main/test/codegen

Also, you need to ensure that the code generated for each and every test case actually works in your local setup. This will help you fix any issue.

**Comment by Clasherzz** [comment]
I will start on that. In the meantime Is this thrid party library okay to use. The multipart library that you mentioned doesn't work for async requests. There is multipart- async which is not as widely used and the documentation is also poor. Currently there is scarcity of good libraries that works along with new hyper versions.This felt like an easier choice.

**Comment by Clasherzz** [comment]
The testfile is added @ashitaprasad

**Comment by Clasherzz** [comment]
The branc is ready for review @ashitaprasad .

File: test/codegen/rust_hyper_codegen_test.dart [hunk 1]
```diff
@@ -0,0 +1,1122 @@
...
+    transfer.perform().unwrap();
+  }
+
+  let response_body = String::from_utf8_lossy(&data);
+
+  println!("Response body: {}", response_body);
+  println!("Response code: {}", easy.response_code().unwrap());
+}""";
+      expect(
+          codeGen.getCode(CodegenLanguage.rustCurl, requestModelPost3, "https"),
```
Comment (ashitaprasad) [inline test/codegen/rust_hyper_codegen_test.dart:688]: incorrect - CodegenLanguage.rustCurl?

File: test/codegen/rust_hyper_codegen_test.dart [hunk 2]
```diff
@@ -0,0 +1,1122 @@
...
+    transfer.perform().unwrap();
+  }
+
+  let response_body = String::from_utf8_lossy(&data);
+
+  println!("Response body: {}", response_body);
+  println!("Response code: {}", easy.response_code().unwrap());
+}""";
+      expect(
+          codeGen.getCode(CodegenLanguage.rustCurl, requestModelPost9, "https"),
```
Comment (ashitaprasad) [inline test/codegen/rust_hyper_codegen_test.dart:948]: incorrect - CodegenLanguage.rustCurl?

**Comment by ashitaprasad** [comment]
There were two instances of `CodegenLanguage.rustCurl` in the test cases.
As per guidelines, you had to validate each and every test case locally and it looks like you haven't done so.

**Comment by Clasherzz** [comment]
> There were two instances of `CodegenLanguage.rustCurl` in the test cases.
> As per guidelines, you had to validate each and every test case locally and it looks like you haven't done so.

There was a commit mistake while committing. I would add the correct files without delay.

**Comment by ashitaprasad** [comment]
Thanks for the support 🚀

**Comment by Clasherzz** [comment]
Hey i just wanted to let you guys know that this was my first open source contribution. Thank for the support @Tanish2002 . Thanks for taking your time and patience on me @ashitaprasad.

**Comment by ashitaprasad** [comment]
@Clasherzz It is definitely a wonderful achievement and you should celebrate it on social media (LinkedIn, Twitter) and let your friends know about it 🎉🙌

**Pull request merged**